### PR TITLE
ddl: fix reassigned partition id in `truncate table` does not take effect

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 
When we executed `truncate table` multiple times, we found that the assigned `partitionID` did not take effect.

### What is changed and how it works?
* Assign a new partition ID to each partition.
* Assigning a new ID to a partition should be done by changing the pointer instead of assigning it by copying.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: ddl/partition.go [hunk 1]
```diff
@@ -417,3 +417,24 @@ func isRangePartitionColUnsignedBigint(cols []*table.Column, pi *model.Partition
 	}
 	return false
 }
+
+// truncateTableFromReassignPartitionIDs reassign a new partition ids.
+func truncateTableFromReassignPartitionIDs(job *model.Job, t *meta.Meta, tblInfo *model.TableInfo) (ver int64, _ error) {
```
Comment (lamxTyler) [inline ddl/partition.go]: Why return the `ver`? It is not used or assigned in this function.

**Comment by ciscoxll** [comment]
@lamxTyler @zimulala PTAL.

**Comment by ciscoxll** [comment]
@tiancaiamao @crazycs520 PTAL.

File: ddl/partition.go [hunk 2]
```diff
@@ -417,3 +417,24 @@ func isRangePartitionColUnsignedBigint(cols []*table.Column, pi *model.Partition
 	}
 	return false
 }
+
+// truncateTableFromReassignPartitionIDs reassign a new partition ids.
+func truncateTableFromReassignPartitionIDs(job *model.Job, t *meta.Meta, tblInfo *model.TableInfo) error {
+	newDefs := make([]model.PartitionDefinition, 0, len(tblInfo.Partition.Definitions))
```
Comment (crazycs520) [inline ddl/partition.go]: no need to define a new defs slice.
```go
for i := range tblInfo.Partition.Definitions {
	var pid int64
	pid, err = t.GenGlobalID()
	if err != nil {
		job.State = model.JobStateCancelled
		return ver, errors.Trace(err)
	}
	tblInfo.Partition.Definitions[i].ID = pid
}
```

File: ddl/partition.go [hunk 3]
```diff
@@ -417,3 +417,16 @@ func isRangePartitionColUnsignedBigint(cols []*table.Column, pi *model.Partition
...
+
+// truncateTableFromReassignPartitionIDs reassign a new partition ids.
+func truncateTableFromReassignPartitionIDs(job *model.Job, t *meta.Meta, tblInfo *model.TableInfo) error {
+	for i := range tblInfo.Partition.Definitions {
+		pid, err := t.GenGlobalID()
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return errors.Trace(err)
+		}
+		tblInfo.Partition.Definitions[i].ID = pid
```
Comment (tiancaiamao) [inline ddl/partition.go]: So, if an error occur and DDL rollback, `model.TableInfo.Partition` is partially changed, some id is new and the old ?

Comment (ciscoxll) [reply ddl/partition.go]: @tiancaiamao Only one `updateSchemaVersion` was executed in `onTruncateTable`, so if an error occurs and DDL rollback will not be a problem.

**Pull request merged**